File: dotnet/src/Experimental/Process.Runtime.Dapr/DaprProcessInfo.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Collections.Generic;
using System.Runtime.Serialization;

namespace Microsoft.SemanticKernel;

/// <summary>
/// A serializable representation of a Dapr Process.
/// </summary>
[KnownType(typeof(KernelProcessEdge))]
[KnownType(typeof(KernelProcessState))]
[KnownType(typeof(KernelProcessMapState))]
[KnownType(typeof(KernelProcessStepState))]
[KnownType(typeof(KernelProcessStepState<>))]
public sealed record DaprProcessInfo : DaprStepInfo
{
    /// <summary>
    /// The collection of Steps in the Process.
    /// </summary>
    public required IList<DaprStepInfo> Steps { get; init; }

    /// <summary>
    /// Initializes a new instance of the <see cref="KernelProcess"/> class from this instance of <see cref="DaprProcessInfo"/>.
    /// </summary>
    /// <returns>An instance of <see cref="KernelProcess"/></returns>
    /// <exception cref="KernelException"></exception>
    public KernelProcess ToKernelProcess()
    {
        var processStepInfo = this.ToKernelProcessStepInfo();
        if (this.State is not KernelProcessState state)
        {
            throw new KernelException($"Unable to read state from process with name '{this.State.Name}' and Id '{this.State.Id}'.");
        }

        List<KernelProcessStepInfo> steps = [];
        foreach (var step in this.Steps)
        {
            if (step is DaprProcessInfo processStep)
            {
                steps.Add(processStep.ToKernelProcess());
            }
            else if (step is DaprMapInfo mapStep)
            {
                steps.Add(mapStep.ToKernelProcessMap());
            }
            else if (step is DaprProxyInfo proxyStep)
            {
                steps.Add(proxyStep.ToKernelProcessProxy());
            }
            else
            {
                steps.Add(step.ToKernelProcessStepInfo());
            }
        }

        return new KernelProcess(state, steps, this.Edges);
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="DaprProcessInfo"/> class from an instance of <see cref="KernelProcess"/>.
    /// </summary>
    /// <param name="kernelProcess">The <see cref="KernelProcess"/> used to build the <see cref="DaprProcessInfo"/></param>
    /// <returns>An instance of <see cref="DaprProcessInfo"/></returns>
    public static DaprProcessInfo FromKernelProcess(KernelProcess kernelProcess)
    {
        Verify.NotNull(kernelProcess);

        DaprStepInfo daprStepInfo = DaprStepInfo.FromKernelStepInfo(kernelProcess);
        List<DaprStepInfo> daprSteps = [];

        foreach (var step in kernelProcess.Steps)
        {
            if (step is KernelProcess processStep)
            {
                daprSteps.Add(DaprProcessInfo.FromKernelProcess(processStep));
            }
            else if (step is KernelProcessMap mapStep)
            {
                daprSteps.Add(DaprMapInfo.FromKernelProcessMap(mapStep));
            }
            else if (step is KernelProcessProxy proxyStep)
            {
                daprSteps.Add(DaprProxyInfo.FromKernelProxyInfo(proxyStep));
            }
            else
            {
                daprSteps.Add(DaprStepInfo.FromKernelStepInfo(step));
            }
        }

        return new DaprProcessInfo
        {
            InnerStepDotnetType = daprStepInfo.InnerStepDotnetType,
            State = daprStepInfo.State,
            Edges = daprStepInfo.Edges,
            Steps = daprSteps,
        };
    }
}

File: python/semantic_kernel/processes/dapr_runtime/dapr_process_info.py
# Copyright (c) Microsoft. All rights reserved.


from typing import Literal

from pydantic import Field

from semantic_kernel.processes.dapr_runtime.dapr_step_info import DaprStepInfo
from semantic_kernel.processes.kernel_process.kernel_process import KernelProcess
from semantic_kernel.processes.kernel_process.kernel_process_state import KernelProcessState
from semantic_kernel.processes.kernel_process.kernel_process_step_info import KernelProcessStepInfo
from semantic_kernel.utils.experimental_decorator import experimental_class


@experimental_class
class DaprProcessInfo(DaprStepInfo):
    """A Dapr process info."""

    type: Literal["DaprProcessInfo"] = Field("DaprProcessInfo")  # type: ignore
    steps: list["DaprStepInfo | DaprProcessInfo"] = Field(default_factory=list)

    def to_kernel_process(self) -> KernelProcess:
        """Converts the Dapr process info to a kernel process."""
        if not isinstance(self.state, KernelProcessState):
            raise ValueError("State must be a kernel process state")

        steps: list[KernelProcessStepInfo] = []
        for step in self.steps:
            if isinstance(step, DaprProcessInfo):
                steps.append(step.to_kernel_process())
            else:
                steps.append(step.to_kernel_process_step_info())

        return KernelProcess(state=self.state, steps=steps, edges=self.edges)

    @classmethod
    def from_kernel_process(cls, kernel_process: KernelProcess) -> "DaprProcessInfo":
        """Creates a Dapr process info from a kernel process."""
        if kernel_process is None:
            raise ValueError("Kernel process must be provided")

        dapr_step_info = DaprStepInfo.from_kernel_step_info(kernel_process)
        dapr_steps: list[DaprProcessInfo | DaprStepInfo] = []

        for step in kernel_process.steps:
            if isinstance(step, KernelProcess):
                dapr_steps.append(DaprProcessInfo.from_kernel_process(step))
            else:
                dapr_steps.append(DaprStepInfo.from_kernel_step_info(step))

        return DaprProcessInfo(
            inner_step_python_type=dapr_step_info.inner_step_python_type,
            state=dapr_step_info.state,
            edges={key: list(value) for key, value in dapr_step_info.edges.items()},
            steps=dapr_steps,
        )
